File: frontend/src/api/templates.ts
import { apiFetch } from './client';
import type {
  SnapshotFromPeriodResponse,
  TemplateItemCreatePayload,
  TemplateItemRead,
  TemplateItemUpdatePayload,
} from './types';

/**
 * GET /api/v1/template/items
 *
 * Returns all template items (single-tenant — no pagination needed).
 * Frontend groups by category_id locally for the TemplateScreen layout.
 */
export async function listTemplateItems(): Promise<TemplateItemRead[]> {
  return apiFetch<TemplateItemRead[]>('/template/items');
}

/**
 * POST /api/v1/template/items
 *
 * Creates a new template-item. Backend validates that category exists and is
 * not archived (D-36); raises 400 otherwise.
 */
export async function createTemplateItem(
  payload: TemplateItemCreatePayload,
): Promise<TemplateItemRead> {
  return apiFetch<TemplateItemRead>('/template/items', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

/**
 * PATCH /api/v1/template/items/{id} — partial update.
 *
 * Pass only the fields to change. Used by inline-edit (amount only) and by
 * the BottomSheet full editor (any subset).
 */
export async function updateTemplateItem(
  id: number,
  patch: TemplateItemUpdatePayload,
): Promise<TemplateItemRead> {
  return apiFetch<TemplateItemRead>(`/template/items/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(patch),
  });
}

/**
 * DELETE /api/v1/template/items/{id} — hard delete.
 *
 * Per CLAUDE.md convention: "Soft delete только для category". Template items
 * are removed permanently; backend returns the deleted row for confirmation.
 */
export async function deleteTemplateItem(id: number): Promise<TemplateItemRead> {
  return apiFetch<TemplateItemRead>(`/template/items/${id}`, { method: 'DELETE' });
}

/**
 * POST /api/v1/template/snapshot-from-period/{period_id} (TPL-03).
 *
 * Destructive overwrite: replaces the entire template with rows derived from
 * the given period's planned-transactions (excluding subscription_auto rows
 * per D-32). Returns the new template items + count of previously-existing
 * rows that were replaced.
 */
export async function snapshotFromPeriod(
  periodId: number,
): Promise<SnapshotFromPeriodResponse> {
  return apiFetch<SnapshotFromPeriodResponse>(
    `/template/snapshot-from-period/${periodId}`,
    { method: 'POST' },
  );
}

File: app/api/schemas/templates.py
"""Pydantic schemas for /api/v1/template/* endpoints (TPL-01, TPL-03)."""
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class TemplateItemCreate(BaseModel):
    """POST /template/items request body."""

    category_id: int = Field(gt=0)
    amount_cents: int = Field(gt=0, description="Amount in kopecks; must be > 0")
    description: Optional[str] = Field(default=None, max_length=500)
    day_of_period: Optional[int] = Field(default=None, ge=1, le=31)
    sort_order: int = Field(default=0, ge=0)


class TemplateItemUpdate(BaseModel):
    """PATCH /template/items/{id} request body — all fields optional."""

    category_id: Optional[int] = Field(default=None, gt=0)
    amount_cents: Optional[int] = Field(default=None, gt=0)
    description: Optional[str] = Field(default=None, max_length=500)
    day_of_period: Optional[int] = Field(default=None, ge=1, le=31)
    sort_order: Optional[int] = Field(default=None, ge=0)


class TemplateItemRead(BaseModel):
    """GET /template/items response item — also returned by POST/PATCH."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    category_id: int
    amount_cents: int
    description: Optional[str]
    day_of_period: Optional[int]
    sort_order: int


class SnapshotFromPeriodResponse(BaseModel):
    """POST /template/snapshot-from-period/{period_id} response (TPL-03)."""

    model_config = ConfigDict(from_attributes=True)

    template_items: list[TemplateItemRead]
    replaced: int = Field(
        ge=0, description="Number of template items deleted before snapshot"
    )
